vertical-pod-autoscaler/pkg/admission-controller/logic/vpa_resource_handler.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logic
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	vpa_types "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/metrics/admission"
+	"k8s.io/klog"
+)
+
+var (
+	possibleUpdateModes = map[vpa_types.UpdateMode]interface{}{
+		vpa_types.UpdateModeOff:      struct{}{},
+		vpa_types.UpdateModeInitial:  struct{}{},
+		vpa_types.UpdateModeRecreate: struct{}{},
+		vpa_types.UpdateModeAuto:     struct{}{},
+	}
+
+	possibleScalingModes = map[vpa_types.ContainerScalingMode]interface{}{
+		vpa_types.ContainerScalingModeAuto: struct{}{},
+		vpa_types.ContainerScalingModeOff:  struct{}{},
+	}
+)
+
+// VpaPreProcessor processes the VPAs before applying default .
+type VpaPreProcessor interface {
+	Process(vpa *vpa_types.VerticalPodAutoscaler, isCreate bool) (*vpa_types.VerticalPodAutoscaler, error)
+}
+
+// noopVpaPreProcessor leaves pods unchanged when processing
+type noopVpaPreProcessor struct{}
+
+// Process leaves the pod unchanged
+func (p *noopVpaPreProcessor) Process(vpa *vpa_types.VerticalPodAutoscaler, isCreate bool) (*vpa_types.VerticalPodAutoscaler, error) {
+	return vpa, nil
+}
+
+// NewDefaultVpaPreProcessor creates a VpaPreProcessor that leaves VPAs unchanged and returns no error
+func NewDefaultVpaPreProcessor() VpaPreProcessor {
+	return &noopVpaPreProcessor{}
+}
+
+// VpaResourceHandler builds patches for VPAs
+type VpaResourceHandler struct {
```

Comment:
You can make `VpaResourceHandler` private and add
```
func NewVpaResourceHandler(vpaPreProcessor VpaPreProcessor) ResourceHandler {
    return &VpaResourceHandler{
        vpaPreProcessor: vpaPreProcessor,
    }
}
```
Now we know that this implements  ResourceHandler interface.
Same change can be done for PodResourceHandler

Reply:
Done.